internal/dag/builder.go

```diff
@@ -917,29 +919,94 @@ func (b *Builder) processTCPProxy(sw *ObjectStatusWriter, ir *ingressroutev1.Ing
 
 		// follow the link and process the target ingress route
 		sw, commit := sw.WithObject(dest)
-		b.processTCPProxy(sw, dest, visited, host)
+		b.processIngressRouteTCPProxy(sw, dest, visited, host)
 		commit()
 	}
 }
 
-func (b *Builder) processTCPProxyHTTPProxy(sw *ObjectStatusWriter, httpproxy *projcontour.HTTPProxy, host string) {
-	if len(httpproxy.Spec.TCPProxy.Services) > 0 {
+// processHTTPProxyTCPProxy processes the spec.tcpproxy stanza in a HTTPProxy document
+// following the chain of spec.tcpproxy.include references. It returns true if processing
+// was successful, otherwise false if an error was encountered. The details of the error
+// will be recorded on the status of the relevant HTTPProxy object,
+func (b *Builder) processHTTPProxyTCPProxy(sw *ObjectStatusWriter, httpproxy *projcontour.HTTPProxy, visited []*projcontour.HTTPProxy, host string) bool {
+	tcpproxy := httpproxy.Spec.TCPProxy
+	if tcpproxy == nil {
+		// nothing to do
+		return true
+	}
+
+	visited = append(visited, httpproxy)
+
+	if len(tcpproxy.Services) > 0 && tcpproxy.Include != nil {
+		sw.SetInvalid("tcpproxy: cannot specify services and include in the same httpproxy")
+		return false
+	}
+
+	if len(tcpproxy.Services) > 0 {
 		var proxy TCPProxy
 		for _, service := range httpproxy.Spec.TCPProxy.Services {
 			m := Meta{name: service.Name, namespace: httpproxy.Namespace}
 			s := b.lookupService(m, intstr.FromInt(service.Port))
 			if s == nil {
 				sw.SetInvalid(fmt.Sprintf("tcpproxy: service %s/%s/%d: not found", httpproxy.Namespace, service.Name, service.Port))
-				return
+				return false
 			}
 			proxy.Clusters = append(proxy.Clusters, &Cluster{
 				Upstream:           s,
-				LoadBalancerPolicy: loadBalancerPolicy(httpproxy.Spec.TCPProxy.LoadBalancerPolicy),
+				LoadBalancerPolicy: loadBalancerPolicy(tcpproxy.LoadBalancerPolicy),
 			})
 		}
 		b.lookupSecureVirtualHost(host).TCPProxy = &proxy
```

Comment:
Nit: I think it's worth a comment there that the existing `TCPProxy` field must be nil because we constrain the tcpproxy spec to only have either services or an include (but not both).

Reply:
Could you say more please, I don't think I understand.

Reply:
Theoretically, `b.lookupSecureVirtualHost(host).TCPProxy` could already have been assigned, in which case this code overwrites it with a different proxy object. I convinced myself that this can't happen in practice because the TCP proxy can only either include or have services (not both). If it was already assigned we must be including something with services after already processing some services, which could only happen if we broke the above rule.

Reply:
Thanks for clarifying (I wanted to check before I went off on a tangent). 

We have a few compensating controls that should ensure that line 959 never overwrites that value. Earlier on we check that an FQDN is unique across all IngressRoute and httpproxy records. If that check holds true then each root httpproxy should only be processed once (roots cannot include other roots) so this condition would hold true.

But I agree it is subtle, and under tested. I've raised #1777 to capture this.